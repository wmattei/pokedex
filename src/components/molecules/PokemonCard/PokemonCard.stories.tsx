import { object, withKnobs } from '@storybook/addon-knobs';
import { storiesOf } from '@storybook/react-native';
import * as React from 'react';
import PokemonCard from '.';
import CenterView from '../../../../storybook/stories/CenterView';

storiesOf('PokemonCard', module)
  .addDecorator((getStory) => <CenterView>{getStory()}</CenterView>)
  .addDecorator(withKnobs)
  .add('Pokemon com imagem', () => (
    <PokemonCard pokemon={object('bulbasaur', { name: 'Bulbassur', id: 1 })} />
  ));
