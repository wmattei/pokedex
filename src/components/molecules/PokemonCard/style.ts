import { StyleSheet } from 'react-native';

export default StyleSheet.create({
  surface: {
    borderRadius: 10,
    width: 160,
    height: 160,
    marginTop: 4,
    marginBottom: 4,
    alignSelf: 'center',
    alignItems: 'center',
    justifyContent: 'center',
  },
  text: {
    marginTop: 10,
    fontSize: 24,
  },
});
