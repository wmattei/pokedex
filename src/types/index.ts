export type KeyboardTypes =
  | 'email-address'
  | 'visible-password'
  | 'default'
  | 'numeric'
  | 'phone-pad'
  | 'number-pad'
  | 'decimal-pad'
  | 'ascii-capable'
  | 'numbers-and-punctuation'
  | 'url'
  | 'name-phone-pad'
  | 'twitter'
  | 'web-search'
  | undefined;

export type TextContentType =
  | 'none'
  | 'URL'
  | 'addressCity'
  | 'addressCityAndState'
  | 'addressState'
  | 'countryName'
  | 'creditCardNumber'
  | 'emailAddress'
  | 'familyName'
  | 'fullStreetAddress'
  | 'givenName'
  | 'jobTitle'
  | 'location'
  | 'middleName'
  | 'name'
  | 'namePrefix'
  | 'nameSuffix'
  | 'nickname'
  | 'organizationName'
  | 'postalCode'
  | 'streetAddressLine1'
  | 'streetAddressLine2'
  | 'sublocality'
  | 'telephoneNumber'
  | 'username'
  | 'password'
  | 'newPassword'
  | 'oneTimeCode';
