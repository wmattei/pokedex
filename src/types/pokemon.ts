export interface IPokemon {
  name: string;
  id: number;
  types?: Array<string>;
  stats?: IPokemonStats;
}

export interface IPokemonStats {
  hp?: number;
  attack?: number;
  defense?: number;
  specialAttack?: number;
  specialDefense?: number;
  speed?: number;
}
