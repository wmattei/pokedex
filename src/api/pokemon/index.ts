import axios from 'axios';
import { IPokemon } from '../../types/pokemon';

type ListPokemonsResponse = {
  count: number;
  next: string;
  previous: string;
  results: Array<IPokemon>;
};

const extractIdFromUrl = (pokemon: { name: string; url: string }) => {
  const paths = pokemon.url.split('/');
  return {
    ...pokemon,
    id: +paths[paths.length - 2],
  };
};

export class PokemonApi {
  static listPokemons(url: string): Promise<ListPokemonsResponse> {
    return axios.get(url).then((res) => {
      return {
        ...res.data,
        results: res.data.results.map(extractIdFromUrl),
      };
    });
  }
}
