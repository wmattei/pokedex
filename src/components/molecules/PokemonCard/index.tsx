import * as React from 'react';
import { Image, Text } from 'react-native';
import { Surface } from 'react-native-paper';
import { POKEMON_IMG_URI } from '../../../constants';
import { IPokemon } from '../../../types/pokemon';
import { FileUtils } from '../../../utils/file';
import { StringUtils } from '../../../utils/string/index';
import style from './style';
// eslint-disable-next-line @typescript-eslint/no-var-requires
const placeholder = require('../../../../assets/pokemon-placeholder.png');

type Props = {
  pokemon: IPokemon;
};

export default React.memo(function PokemonCard({ pokemon }: Props) {
  const [imageSrc, setImageSrc] = React.useState('');
  React.useEffect(() => {
    FileUtils.fetchImageFromCache(`${POKEMON_IMG_URI}/${pokemon.id}.png`).then(
      setImageSrc
    );
  }, []);

  return (
    <Surface style={style.surface}>
      <Image
        style={{ width: 100, height: 100 }}
        source={imageSrc ? { uri: imageSrc } : placeholder}
      />
      <Text style={style.text}>{StringUtils.toCapitalCase(pokemon.name)}</Text>
    </Surface>
  );
});
