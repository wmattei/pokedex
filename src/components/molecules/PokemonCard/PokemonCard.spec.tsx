import * as React from 'react';
import { create } from 'react-test-renderer';
import PokemonCard from '.';
import { IPokemon } from '../../../types/pokemon';

describe('PokemonCard', () => {
  it('Renders correctly', () => {
    const pokemon: IPokemon = {
      id: 1,
      name: 'Bulbasaur',
    };
    const tree = create(<PokemonCard pokemon={pokemon} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
