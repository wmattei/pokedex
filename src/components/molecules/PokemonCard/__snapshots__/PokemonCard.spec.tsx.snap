// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonCard Renders correctly 1`] = `
<View
  accessible={true}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      false,
      undefined,
    ]
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "alignSelf": "center",
        "backgroundColor": "#ffffff",
        "borderRadius": 10,
        "height": 160,
        "justifyContent": "center",
        "marginBottom": 4,
        "marginTop": 4,
        "shadowColor": "#000000",
        "shadowOffset": Object {
          "height": 3,
          "width": 0,
        },
        "shadowOpacity": 0.24,
        "shadowRadius": 4,
        "width": 160,
      }
    }
  >
    <View
      style={
        Object {
          "height": 100,
          "width": 100,
        }
      }
    >
      <Image
        onError={[Function]}
        source={
          Object {
            "uri": "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAOEAAADhCAMAAAAJbSJIAAAAUVBMVEX///8AAAD09PTMzMzZ2dkhISHv7+/m5ubU1NRKSkrHx8eOjo5QUFCpqal4eHhVVVUrKyv4+PiysrKqqqpBQUEZGRk1NTU7OzsJCQlzc3OUlJQZO629AAABVElEQVR4nO3cy07CUBSGUS6VigoqN9H3f1BHpZOdyC2esw9rjUvyfzkMmjbpZAIAAAAAAAAAAAAAAAAw6Obn6ktPvdJyeq5V6alXeleosHoKFdZP4SMVbkpPvUS3exvsTwW7PtCNhYfTb75K7//bc3RGy+jKWXic/733cgpHCmulcKSwVgpHCmulcKSwVgpHCmulcKSwVmFh+PQla+HsJTAPr0xaeD6F+SnMT2F+CvNrtXARvV3bfwwaKFxPj4OxcDX5HJTed7t19Nfcll51T2Fh1vf4IYX5KcxPYX6NFvbf68HhlLXfbgZd6YE366KDe2rgDu0kLpyVnnVHCvNTmJ/C/BTmpzA/hfkpzE9hfg9aeGypsH8N/LT0nAYAoAUPeuetMBWF+SnMT2F+CvNTmJ/C/BTmpzC/9gvDDy1MWyqcLCKlRwEAAAAAAAAAAAAAAAB1+gWlkhf/xP9IqQAAAABJRU5ErkJggg==",
          }
        }
        style={
          Array [
            Object {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            },
            Object {},
          ]
        }
        transitionDuration={300}
        uri="https://pokeres.bastionbot.org/images/pokemon//1.png"
      />
    </View>
    <Text
      style={
        Object {
          "fontSize": 24,
          "marginTop": 10,
        }
      }
    >
      Bulbasaur
    </Text>
  </View>
</View>
`;
